jax/lax/lax_control_flow.py

```diff
@@ -888,65 +893,81 @@ def _cond_partial_eval(trace, *tracers, branches, linear):
 # staged jaxpr that accepts those residuals as its first few inputs. The
 # residual-producing branches are staged as jaxprs and bound right away in a
 # conditional. The residual-consuming jaxprs are assembled together in a jaxpr
-# conditional. The following two helper functions are used to ensure that both
-# collections of jaxprs (those evaluated and those staged) are valid for joint
-# use under their respective conditionals.
-
-# Because every branch might produce different residuals, the branches' output
-# signatures might not match. But we need branch signatures to match in order to
-# bind them in a conditional. This function "joins" the residual outputs of the
-# branches by concatenation. Each augmented branch returns zero-filled values in
-# the place of all other branches' residuals.
-def _join_cond_outputs(jaxprs, res_avals_per_jaxpr, num_non_res_outputs):
-  def augment_jaxpr(i, jaxpr):
-    res_avals_prefix = util.concatenate(res_avals_per_jaxpr[:i])
-    res_avals_suffix = util.concatenate(res_avals_per_jaxpr[i+1:])
-
+# conditional. The following helper functions ensure that both collections of
+# jaxprs (those evaluated and those staged) are valid for joint use under their
+# respective conditionals.
+#
+# In particular, the residuals derived from each original branch may have
+# distinct types. Because the branches of conditionals must have identical type
+# signatures, we join residuals together across branches into a common format.
+
+# In order to set up a type signature that all branches can conform to, it would
+# suffice to concatenate all branches' residuals. But concatenation can result
+# in redundant inputs and outputs, and might result in memory allocation that
+# scales unnecessarily with the branch count. This function finds common
+# residual types across branches for reuse, so as to avoid redundant allocation.
+def _merge_branch_residuals(branch_res_avals):
+  def enumerate_equal(xs):
+    counts = {v: itertools.count() for v in set(xs)}
+    return [(x, next(counts[x])) for x in xs]
+  branch_res_avals = _map(enumerate_equal, branch_res_avals)
+  all_avals = sorted(set(util.concatenate(branch_res_avals)))
```

Comment:
I want to check my understanding of this logic. I think `util.concatenate(branch_res_avals)` is a list of (aval, int) pairs, where the int represents a tag so that in effect `enumerate_equal` is returning a disjoint union of abstract values (but we preserve the order by returning a list rather than a set, since we use this order in what comes next). We then do a union of these disjoint unions on this line.

I'm not sure why this `sorted` seems to work, since abstract values can't be compared. I was surprised that  `(None, 3) < (None, 4)` works even though `None < None` is an error, but [the Python docs say](https://docs.python.org/3/reference/expressions.html#value-comparisons) "The built-in containers typically assume identical objects are equal to themselves. That lets them bypass equality tests for identical objects to improve performance and to maintain their internal invariants." Could we be relying on that here, or is something else allowing this set of pairs to be sorted? Do we need this to be sorted at all? LMK if I'm missing something!

Reply:
You're correct in both. I used `sorted` in an attempt to ensure determinism across invocations (as I understand it, set iteration order is not specified). I'd only like to have a consistent ordering, and for that we don't need `sorted` regardless of whether or how it works. Fixed.